[Excel] (Custom Functions) Custom functions architecture article (#805)
OfficeDev/office-js-docs-pr: docs/excel/custom-functions-architecture.md

---
ms.date: 03/06/2019
description: Learn about Excel custom functions' runtime. 
title: Custom functions architecture (preview)
localization_priority: Priority
---
# Custom functions architecture

> Custom function use a new JavaScript runtime that differs from the browser-based JavaScript engine which powers most other parts of your add-in.

Reviewer: Would be good to add a stronger "why keep reading" or intent to the first sentence. 
- Should mention there are some  that get talked about in this article. (like you can't load any DOM usage libraries and you can't directly communicate with task pane code, etc...)
- I wouldn't call it a "new" runtime. It's just a separate runtime for reasons outlined below.

## Custom functions runtime

> The custom functions runtime prioritizes executing calculations, so functions run run smoothly when Excel recalculates. The custom functions runtime allows access to APIs which make requesting external data and exchanging data over a persistent connection over a server possible.

Reviewer: I'm not clear on the second sentence. Is it referring to APIs in the JS library that enable requesting external data?

> The custom functions runtime prioritizes executing calculations, so functions run run smoothly when Excel recalculates. The custom functions runtime allows access to APIs which make requesting external data and exchanging data over a persistent connection over a server possible.

Author: Yeah - because it also allows you to get data from your own server via websockets it seemed weird to just say "it can also do web things!" but happy to rephrase.

> Because functions are limited to actions performed in a workbook, the custom functions runtime doesn't support UI elements nor does it provide access to the Document Object Model (DOM). You should think of custom functions as a separate component of your add-in, separate from your add-in's task pane or other UI elements.

Reviewer: "limited to actions performed in a workbook" sort of implies you can have UI since task panes also perform actions in a workbook like UI interactions. Could be clearer to say that custom functions purpose is to return a calculated value for a cell and nothing else.

Reviewer: I wonder if we should have a bullet list of these items. At any rate, we should be more clear about what it means to not have a DOM. for example, you cannot use JQuery or any other libraries that would rely on a DOM.

Also we should mention that you cannot use local storage to share data with task panes or other sections of code. Each runtime has its own local storage. (Maybe this would be good to mention in AsyncStorage later just to help prevent developers from accidentally thinking they can use local storage.)

> Because functions are limited to actions performed in a workbook, the custom functions runtime doesn't support UI elements nor does it provide access to the Document Object Model (DOM). You should think of custom functions as a separate component of your add-in, separate from your add-in's task pane or other UI elements.

Reviewer: ```suggestion
Because custom functions are limited to actions performed in a workbook, the custom functions runtime doesn't support UI elements nor does it provide access to the Document Object Model (DOM). You can think of custom functions as a separate component of your add-in, separate from your add-in's task pane or other UI elements.
```

> >[!NOTE] 

Reviewer: double-check but I don't think this is the case anymore. You always get a task pane now with the new templates.

or thinking about this more, maybe we should document the html file that is the loader for your custom functions. Just to point out its purpose and to not use it as your task pane (task panes are under the task pane folder).

> When first installing a custom function, a task pane is shown. This pane is necessary to install the custom functions runtime, but shouldn't serve as your main task pane for your add-in.

> ## Browser-based engine

Reviewer: I thought we decided on "Browser engine runtime"?

should make this consistent throughout

> You may be used to the common browser-based engine which powers most add-ins if you have developed an add-in before. This engine allows access to the Office.js APIs. Any of the Excel APIs (such as APIs which allow you to manipulate Excel tables, charts, etc) you may have used when developing an Excel add-in run on this browser-based engine, but aren't directly accessible from the custom functions runtime. 

Reviewer: I would just phrase more directly as "The task pane and other components outside of custom functions run in a browser engine runtime. The browser engine runtime enables access to the Office.js APIs..."

The browser-based engine is also what supports the task pane and UI elements of your add-in. Keep this in mind while developing an add-in which utilizes both a task pane and a custom function, especially if those two parts of your add-in need to communicate.

> ## Communicate between engines

Reviewer: ```suggestion
## Communicate between runtimes
```

> Both the browser-based engine and custom functions runtime have access to new APIs in the `OfficeRuntime` namespace. Use these APIs to facilitate passing information between your function and the UI elements of your add-in. `OfficeRuntime` APIs offer a storage location (`AsyncStorage`) and a new dialog API which is compatible with custom functions.

Reviewer: I would just shorten this up. Really just need to point out AsyncStorage. 

Like "Both the browser-based engine and custom functions runtime have AsyncStorage APIs to share data... You can store data into AsyncStorage from your custom function code and get the data from your task pane code..."

## Sample scenarios

> `AsyncStorage` is helpful in almost every situation in which you wish to use Office.js APIs with a custom function. One common scenario is authenticating users through the task pane before they are given access to a custom function. You'll find a code sample of this authentication scenario in this [Github repository](https://github.com/OfficeDev/PnP-OfficeAddins/tree/master/Excel-custom-functions/AsyncStorage) dedicated to patterns and practices.

Reviewer: This code sample just shows how to pass data in AsyncStorage, it does not show end to end auth. Just should clarify.

---
ms.date: 03/20/2019
description: Learn about Excel custom functions' runtime. 
title: Custom functions architecture (preview)
localization_priority: Priority
---
# Custom functions architecture

 Custom functions are with their own unique runtime that prioritizes execution of calculations. This article will cover the differences between the custom functions runtime and the browser-based JavaScript engine which powers most other parts of your add-in.

## Custom functions runtime

An Office Web Add-in can interact with the user as a task pane, or a content pane, and can include commands and custom functions. All of these parts run in a browser engine runtime except for custom functions. Custom functions run in a separate custom functions runtime to optimize for calculation speed.

> Note that if you are using the [Yeoman generator for Office Add-ins](https://www.npmjs.com/package/generator-office) to generate your project, the custom functions runtime will be loaded through the custom-functions.js script file referenced in the functions.html file. The functions.html serves only to load the runtime and shouldn't be used as the task pane for your add-in.

Reviewer 2: ```suggestion
Note that if you're using the [Yeoman generator for Office Add-ins](https://www.npmjs.com/package/generator-office) to generate your project, the custom functions runtime will load through the custom-functions.js script file referenced in the functions.html file. The functions.html serves only to load the runtime and shouldn't be used as the task pane for your add-in.
```

The following table highlights the differences between the custom functions runtime and the browser engine runtime:

| Custom functions runtime 	| Browser engine runtime 	|
|------------------------------------------------------------------	|--------------------------------------------------------------------------------------------------------------	|
| Supports returning a value from a cell 	| Supports Office.js APIs and UI elements 	|
| Does not have `localStorage` object, instead uses `AsyncStorage` 	| Has `localStorage` object, can optionally use `AsyncStorage` object 	|
| Does not support interacting with the DOM, or loading libraries that depend on the DOM such as jQuery.	| Supports interacting with the DOM and loading libraries that depend on the DOM. |

## Browser engine runtime

The task pane, content add-in, and commands run in a browser engine runtime.

The browser engine runtime supports the Office.js APIs. Keep in mind that any of the Excel APIs, such as APIs which allow you to manipulate Excel tables, run on the browser engine runtime, but aren't directly accessible from the custom functions runtime.

## Communicate between runtimes

> Code in your custom functions cannot directly interact with code in other parts of your web add-in, like the task pane because they are in different runtimes. But in some scenarios you may need to share data, such as passing a token.

Reviewer 2: ```suggestion
Your custom functions code cannot directly interact with code in other parts of your web add-in, like the task pane because they are in different runtimes. But in some scenarios you may need to share data, such as passing a token.
```

---
ms.date: 03/20/2019
description: Learn about Excel custom functions' runtime. 
title: Custom functions architecture (preview)
localization_priority: Priority
---
# Custom functions architecture

 Custom functions are with their own unique runtime that prioritizes execution of calculations. This article will cover the differences between the custom functions runtime and the browser-based JavaScript engine which powers most other parts of your add-in.

## Custom functions runtime

> Because of their unique runtime, think of custom functions as another component of your total add-in. Your add-in will also likely include elements from the browser engine runtime, like a task pane.

Reviewer: Maybe we should just list all the parts to make this clearer. Would make a good picture actually to show visually how it all runs (perhaps later).

"An Office Web Add-in can interact with the user as a task pane, or a content pane, and can include commands and custom functions. All of these parts run in a browser engine runtime except for custom functions. Custom functions run in a separate custom functions runtime to optimize for calculation speed..." 

or something like that.

> The following table highlights the differences between the custom functions runtime and the browser engine runtime:

Reviewer: I like this table!

| Custom functions runtime 	| Browser engine runtime 	|
|------------------------------------------------------------------	|--------------------------------------------------------------------------------------------------------------	|
| Supports returning a value from a cell 	| Supports Office.js APIs and UI elements 	|
| Does not have `localStorage` object, instead uses `AsyncStorage` 	| Has `localStorage` object, can optionally use `AsyncStorage` object 	|

> | Provides access to `OfficeRuntime` objects 	| Provides access to `OfficeRuntime` objects 	|

Reviewer: I think I would leave OfficeRuntime off the list, or explain it differently. OfficeRuntime is on Office.js side, but only for AsyncStorage as far as I know. That's sort of covered already in the previous row.

> | Allows you to make web requests 	| Allows you to access the Document Object Model (DOM) and supports libraries that use the DOM, such as jQuery 	|

Reviewer: The comparison are not quite equal. Making web requests is supported by both runtimes. In the case of DOM this is a true difference. You would need to state that custom functions cannot load libraries that require DOM, such as jquery.

"| Does not support interacting with the DOM, or loading libraries that depend on the DOM such as jQuery. | Supports interacting with the DOM, and loading libraries that depend on the DOM."

> >[!NOTE]

Reviewer: I would remove this note. Adam updated the template so you will always have a true task pane. However, this article is an excellent space to document the functions.html which loads custom-functions-runtime.js. This is the loader page to load your custom functions into the runtime and make the runtime JS available.

> When first installing a custom function, a task pane is shown. This pane is necessary to install the custom functions runtime, but shouldn't serve as the task pane for your add-in.

## Browser engine runtime

> The task pane and other components that are not custom functions run in a browser engine runtime.

Reviewer: ```suggestion
The task pane, content add-in, and commands run in a browser engine runtime.
```

I think we should just identify them all so there is no ambiguity (especially for anyone new).

> The browser engine runtime enables access to the Office.js APIs. Keep in mind that any of the Excel APIs, such as APIs which allow you to manipulate Excel tables, run on the browser engine runtime, but aren't directly accessible from the custom functions runtime.

Reviewer: ```suggestion
The browser engine runtime supports the Office.js APIs. Keep in mind that the Excel APIs in Office.js, such as APIs which allow you to manipulate Excel tables, run on the browser engine runtime. Office.js APIs are not available from the custom functions runtime.
```
I think technically the custom function APIs are also "Excel APIs" so just making this more specific.

## Communicate between runtimes

> Both the browser engine runtime and the custom functions runtime have access to `OfficeRuntime.AsyncStorage` and `OfficeRuntime.Dialog` APIs.

Reviewer: Maybe start off with a scenario to give this more of a why.

suggestion...
"Code in your custom functions cannot directly interact with code in other parts of your web add-in, like the task pane because they are in different runtimes. But in some scenarios you may need to share data, such as passing a token. AsyncStorage can be used to store data..."

`AsyncStorage` can be used to store data from your custom functions and get data from your task pane code.

> You'll find a code sample using `AsyncStorage` in this [Github repository](https://github.com/OfficeDev/PnP-OfficeAddins/tree/master/Excel-custom-functions/AsyncStorage) dedicated to patterns and practices.

Reviewer: ```suggestion
You can see an example of using `AsyncStorage` in the [Office Add-ins PnP Github repository](https://github.com/OfficeDev/PnP-OfficeAddins/tree/master/Excel-custom-functions/AsyncStorage).
```
